### PR TITLE
JAVACLI-125 Provide more descriptive error on 403

### PR DESCRIPTION
**Pull request by berkeleysquare**
Try to use common exception handling to provide commonality in messages and a single place to test for certain codes.

### CODE REVIEW

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/util/Utils.java [hunk 1]
```diff
@@ -49,27 +51,31 @@
...
+            for (int i = 0; i < versionInfo.length && i < buildInfoArr.length; i++) {
+                final int i1 = Integer.parseInt(buildInfoArr[i]);
+                final int i2 = Integer.parseInt(versionInfo[i]);
+                if (i1 > i2) return true;
+                if (i1 < i2) return false;
+            }
 
-        return true;
+            return true;
+        } catch (IOException e) {
```
Comment (RachelTucker) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/util/Utils.java]: missing final

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/exceptions/CommandExceptionFactory.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+    /**
+     * Provide an exception with a common message "[Command class] failed: reason, then exception
+     *
+     * @param commandName pass in simple name of command class or string describing module
+     * @param e the caught exception
+     * @return a new Command Exception with regularized message.
+     */
+
+    public static CommandException getResponseExcepion(final String commandName, final IOException e) {
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/exceptions/CommandExceptionFactory.java]: `Excepion` should be `Exception`

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/exceptions/CommandExceptionFactory.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        return new CommandException(commandName + " failed: " + getExcepionDescription(e), e);
+    }
+
+    /**
+     * Test for common return types and create description
+     * @param e
+     * @return String describing general cause
+     */
+    public static String getExcepionDescription(final IOException e) {
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/exceptions/CommandExceptionFactory.java]: `Excepion` should be `Exception`

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/exceptions/CommandExceptionFactory.java (hunk 2 above)
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/exceptions/CommandExceptionFactory.java]: I don't know that the name of the first argument reflects that is actually a commandName.  You also pass in other exception types.

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/exceptions/CommandExceptionFactory.java [hunk 4]
```diff
@@ -16,49 +16,57 @@
...
-     * Provide an exception with a common message "[Command class] failed: reason, then exception
-     *
-     * @param commandName pass in simple name of command class or string describing module
-     * @param e the caught exception
-     * @return a new Command Exception with regularized message.
+     * Register an exception handler
+     * @param exceptionClass (extends Throwable)
+     * @param handler (implements Ds3ExceptionHandler)
      */
+    public void addHandler(Class<? extends Throwable> exceptionClass, Ds3ExceptionHandler handler) {
```
Comment (RachelTucker) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/exceptions/CommandExceptionFactory.java]: missing finals

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/util/Utils.java [hunk 5]
```diff
@@ -74,7 +74,8 @@ public static boolean isVersionSupported(final Ds3Client client, final String mi
 
             return true;
         } catch (IOException e) {
```
Comment (RachelTucker) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/util/Utils.java]: missing final

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/exceptions/CommandExceptionFactory.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+    }
+
+    /**
+     * Checks exception type == FailedRequestException and status code
+     * @param e caught exception
+     * @param code code to test
+     * @return true if exception is type of FailedRequestException and has status code == code
+     */
+    public static boolean hasStatusCode(final Exception e, final int code) {
```
Comment (berkeleysquare) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/exceptions/CommandExceptionFactory.java]: This gives the convenience of calling:
            if (CommandExceptionFactory.hasStatusCode(e,404)) {

To provide more information. In getUser it says "Unknown User 'xxxx'" on a 404. There are a smattering of those already in the code base and I hate to rip them out and provide less info.

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/command/GetObjectsOnTape.java [hunk 7]
```diff
@@ -46,23 +47,21 @@ public CliCommand init(final Arguments args) throws Exception {
...
-            if(e.getStatusCode() == 500) {
-                throw new CommandException("Error: Cannot communicate with the remote DS3 appliance.", e);
-            }
-            else if(e.getStatusCode() == 404) {
-                throw new CommandException("Unknown tape: " + this.tapeId, e);
-            }
-            else {
-                throw new CommandException("Encountered an unknown error of ("+ e.getStatusCode() +") while accessing the remote DS3 appliance.", e);
+        } catch (final IOException e) {
+            if (CommandExceptionFactory.hasStatusCode(e, 404)) {
```
Comment (berkeleysquare) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/command/GetObjectsOnTape.java]: This is what I said about this in the factory. You can provide a more tailored message:
GetObjectsOnTape failed (CommandException): Unknown tape '1234'

Instead of 
GetObjectsOnTape failed (CommandException): Entity not found.

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/Main.java [hunk 8]
```diff
@@ -160,15 +168,13 @@ public static void main(final String[] args) {
...
-            LOG.info(e.getResponseString());
+        } catch (final RuntimeException e) {
+            // already processed by exception factory, print and exit
+            System.out.println(e.getMessage());
             System.exit(2);
         } catch (final Exception e) {
-            System.out.println("ERROR: " + e.getMessage());
-            LOG.info("Stack trace: ", e);
+            // run through handler
+            EXCEPTION.handleException(Main.class.getSimpleName(), e, false);
```
Comment (berkeleysquare) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/Main.java]: Agreed -- I put it in a RuntimeHandler

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/command/GetObjectsOnTape.java (hunk 7 above)
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/command/GetObjectsOnTape.java]: So, rather than using the CommandExceptionFactory needing to do a check, let the command just return a different command exception.  The CommandExceptionFactory should not be involved in what exception gets thrown, just what to do when a specific exception is being handled.

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/exceptions/CommandExceptionFactory.java (hunk 6 above)
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/exceptions/CommandExceptionFactory.java]: As I commented earlier, this functionality is not the responsibility of the CommandExceptionFactory (even that name should probably be improved...).  If you do want some kind of utility class to help with that determination, then create that util class.

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/exceptions/Ds3ExceptionHandler.java [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+ *   CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ *   specific language governing permissions and limitations under the License.
+ * ***************************************************************************
+ */
+
+package com.spectralogic.ds3cli.exceptions;
+
+public interface Ds3ExceptionHandler<Throwable> {
+
+    public void handle(String locationDescription, Throwable e, boolean throwRuntimeException);
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/exceptions/Ds3ExceptionHandler.java]: Missing finals here in the interface.

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/exceptions/CommandExceptionFactory.java [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+ * ***************************************************************************
+ */
+
+package com.spectralogic.ds3cli.exceptions;
+
+import com.spectralogic.ds3client.networking.FailedRequestException;
+import java.util.HashMap;
+import java.util.Map;
+
+public class CommandExceptionFactory {
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/exceptions/CommandExceptionFactory.java]: We should update the name of this to better reflect what it is actually responsible for.  Specifically, this is the ExceptionHandlerMapper.  It's not a Factory.

Also, some javadoc to explain what this class is doing/responsible for should be added.

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/exceptions/CommandExceptionFactory.java [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+
+    private static CommandExceptionFactory factoryInstance;
+    private final Map<Class<? extends Throwable>, Ds3ExceptionHandler> handlers = new HashMap();
+    private final Ds3ExceptionHandler defaultHandler = new DefaultExceptionHandler();
+
+    // single instance
+    public static CommandExceptionFactory getInstance() {
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/exceptions/CommandExceptionFactory.java]: If this is a singleton, just create the singleton in the static constructor rather than lazily evaluating it like this.

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/exceptions/Ds3ExceptionHandler.java [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+
+package com.spectralogic.ds3cli.exceptions;
+
+public interface Ds3ExceptionHandler<Throwable> {
+
+    public void handle(final Throwable e);
+
+    // implemeters are encouraged to use format() as the message
+    // to facilitate testing
+    public String format(final Throwable e);
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/exceptions/Ds3ExceptionHandler.java]: I don't think this is needed since it's an internal call and not used by an external interface.

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/exceptions/Ds3ExceptionHandlerFactory.java [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+ */
+
+package com.spectralogic.ds3cli.exceptions;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class Ds3ExceptionHandlerFactory {
+
+    private static Ds3ExceptionHandlerFactory factoryInstance = new Ds3ExceptionHandlerFactory();
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/exceptions/Ds3ExceptionHandlerFactory.java]: This seems like it should be final

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/exceptions/Ds3ExceptionHandlerFactory.java [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+ *   specific language governing permissions and limitations under the License.
+ * ***************************************************************************
+ */
+
+package com.spectralogic.ds3cli.exceptions;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class Ds3ExceptionHandlerFactory {
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/exceptions/Ds3ExceptionHandlerFactory.java]: This is not a factory, remove the name, and instead call it the Ds3ExceptionHandlerMapper

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/exceptions/Ds3ExceptionHandlerFactory.java [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+        handler.handle(e);
+    }
+
+    /**
+     * Locate the appropriate Ds3ExceptionHandler (match if registered, else Default)
+     * then call format()
+     *
+     * @param e
+     */
+    public String formatException (final Throwable e) {
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/exceptions/Ds3ExceptionHandlerFactory.java]: Why is this method needed?

Comment (berkeleysquare) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/exceptions/Ds3ExceptionHandlerFactory.java]: Breaking it into handle() and format() facilitates test -- you can call format() in Unit testing and verify the output.

Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/exceptions/Ds3ExceptionHandlerFactory.java]: My question is more about, why is it needed in the HandlerFactory?  Having different handle from format methods makes sense in individual handlers for testing, but why is this needed at this level of the api?

File: ds3_java_cli/src/test/java/com/spectralogic/ds3cli/Ds3Cli_Test.java [hunk 16]
```diff
@@ -843,9 +838,16 @@ public void getBulkWithBadArgs() throws Exception {
         when(helpers.startReadAllJob(eq("bucketName"), any(ReadJobOptions.class))).thenReturn(mockedGetJob);
 
         final Ds3Cli cli = new Ds3Cli(new Ds3ProviderImpl(null, helpers), args, mockedFileUtils);
-        final CommandResponse result = cli.call();
-        assertThat(result.getMessage(), is(expected));
-        assertThat(result.getReturnCode(), is(1));
+        try {
+            final CommandResponse result = cli.call();
+
+        } catch (Exception e) {
```
Comment (rpmoore) [inline ds3_java_cli/src/test/java/com/spectralogic/ds3cli/Ds3Cli_Test.java]: Missing final

File: ds3_java_cli/src/test/java/com/spectralogic/ds3cli/Ds3Cli_Test.java [hunk 17]
```diff
@@ -1783,8 +1785,16 @@ public void modifyDataPolicyWithBadParam() throws Exception {
         when(client.getDataPolicySpectraS3(any(GetDataPolicySpectraS3Request.class))).thenReturn(GetDataPolicyResponse);
 
         final Ds3Cli cli = new Ds3Cli(new Ds3ProviderImpl(client, null), args, null);
-        final CommandResponse result = cli.call();
-        assertThat(result.getMessage(), is(expected));
+        try {
+            final CommandResponse result = cli.call();
+
+        } catch (CommandException e) {
```
Comment (rpmoore) [inline ds3_java_cli/src/test/java/com/spectralogic/ds3cli/Ds3Cli_Test.java]: Missing final
